wasm/src/lib.rs

```diff
@@ -0,0 +1,188 @@
+    unreachable!("Program should have aborted")
+}
+
+#[cfg(not(test))]
+#[no_mangle]
+#[alloc_error_handler]
+fn oom(layout: ::core::alloc::Layout) -> ! {
+    panic!("Allocation({} bytes) failed", layout.size())
+}
+
+#[no_mangle]
+extern "C" fn _iroha_wasm_alloc(len: u32) -> u32 {
+    core::mem::ManuallyDrop::new(Vec::<u8>::with_capacity(len as usize)).as_mut_ptr() as u32
+}
+
+/// Host exports
+mod host {
+    /// Helper struct which guarantees to be FFI safe since tuple is not
+    #[repr(C)]
+    #[derive(Debug, Clone, Copy)]
+    pub(super) struct WasmQueryResult(pub u32, pub u32);
+
+    #[link(wasm_import_module = "iroha")]
+    extern "C" {
+        /// Executes encoded query by providing offset and length
+        /// into WebAssembly's linear memory where query is stored
+        ///
+        /// # Warning
+        ///
+        /// This function doesn't take ownership of the provided allocation
+        /// but it does transfer ownership of the result to the caller
+        #[cfg(not(test))]
+        pub(super) fn execute_query(ptr: u32, len: u32) -> WasmQueryResult;
+
+        /// Executes encoded instruction by providing offset and length
+        /// into WebAssembly's linear memory where instruction is stored
+        ///
+        /// # Warning
+        ///
+        /// This function doesn't take ownership of the provided allocation
+        /// but it does transfer ownership of the result to the caller
+        #[cfg(not(test))]
+        pub(super) fn execute_instruction(ptr: u32, len: u32);
+    }
+}
+
+/// Decode the object from given pointer and length
+///
+/// # Warning
+///
+/// This method takes ownership of the given pointer
+pub unsafe fn _decode_from_raw<T: Decode>(ptr: u32, len: u32) -> T {
+    let bytes = Box::from_raw(core::slice::from_raw_parts_mut(ptr as *mut _, len as usize));
+
+    T::decode(&mut &bytes[..]).expect(
+        format!(
+            "Decoding of {} failed. This is a bug",
+            core::any::type_name::<T>()
+        )
+        .as_str(),
+    )
+}
+
+/// Encode the given object and call the given function with the pointer and length of the allocation
+///
+/// # Warning
+///
+/// Ownership of the returned allocation is transfered to the caller
+///
+/// # Safety
+///
+/// The given function must not take ownership of the pointer argument
+unsafe fn encode_and_execute<T: Encode, O>(obj: T, fun: unsafe extern "C" fn(u32, u32) -> O) -> O {
+    // NOTE: It's imperative that encoded object is stored on the heap
+    // because heap corresponds to linear memory when compiled to wasm
+    let bytes = obj.encode();
+
+    // NOTE: This is valid as long as it's compiled for 32-bit architecture
+    #[allow(clippy::cast_possible_truncation)]
+    let ptr = bytes.as_ptr() as u32;
+    #[allow(clippy::cast_possible_truncation)]
+    let len = bytes.len() as u32;
+
+    fun(ptr, len)
+}
+
+/// Executes the given query on the host environment
+pub fn execute_query(query: QueryBox) -> QueryResult {
```

Comment:
the same goes for `execute_instruction`

Reply:
I'm not sure, in that it *could* cause a deadlock, maybe bring this up during status with @e-ivkov ?

Reply:
interesting idea but I don't understand how this could deadlock

Reply:
Suppose you only check things that can fail immediately? It can't. The problem arises in, can you guarantee that all checks that fail, do so immediately (which could stall Iroha) or that they don't depend on other instructions in the pipeline? Or that the other instruction in the pipeline doesn't use this as a trigger. 

We could get around this issue, by allowing only a specific kind of error to be returned: one that people aren't tempted to use everywhere, but for specific failure conditions. Kinda like in #1742 ?